File: computer3/Winter Term/9 - Mar 12/lvl3_hw_sol/blocks.cpp
#include <bits/stdc++.h>
using namespace std;
int n; string blk[4], word;
int main(){
    cin >> n;
    for(int i=0; i<4; i++) cin >> blk[i];
    for(int i=0; i<n; i++) {
        cin >> word;
        sort(blk, blk+4);
        bool found = false;
        do{
            bool flag = true;
            for(int k=0; k<word.size(); k++){
                if(blk[k].find(word[k]) == string::npos)
                    flag = false;
            }
            if(flag){ found = true; break;}
        }while(next_permutation(blk, blk+4));
        cout << (found? "YES\n" : "NO\n");
    }
}

File: computer3/Winter Term/9 - Mar 12/lvl3_hw_sol/blocks.py
from itertools import permutations
from sys import stdin
input = stdin.readline
n = int(input())
block = []
for i in range(4):
    block.append(input().strip())
for _ in range(n):
    word = input().strip()
    id = [0, 1, 2, 3]
    found = False
    for x in list(permutations(id)):
        flag = True
        for k in range(len(word)):
            if block[x[k]].find(word[k]) == -1:
                flag = False
                break
        if flag:
            found = True
            break
    print('YES') if found else print('NO')